GitCommands/Submodules/SubmoduleStatusProvider.cs

```diff
@@ -66,77 +65,75 @@ public void Init()
-                }
+            // Structure is updated
+            OnStatusUpdated(result, cancelToken);
 
-                foreach (var info in result.SuperSubmodules)
-                {
-                    _submoduleInfos[info.Path] = info;
-                }
+            // Prepare info for status updates (normally triggered by StatusMonitor)
+            foreach (var info in result.OurSubmodules)
+            {
+                _submoduleInfos[info.Path] = info;
+            }
+
+            foreach (var info in result.SuperSubmodules)
+            {
+                _submoduleInfos[info.Path] = info;
+            }
 
-                if (!_submoduleInfos.ContainsKey(result.TopProject.Path))
+            if (!_submoduleInfos.ContainsKey(result.TopProject.Path))
+            {
+                _submoduleInfos.Add(result.TopProject.Path, result.TopProject);
+            }
+
+            // Start update status for the submodules
+            if (updateStatus)
+            {
+                if (result.SuperProject != null)
                 {
-                    _submoduleInfos.Add(result.TopProject.Path, result.TopProject);
+                    // Update from top module (will stop at current)
+                    await GetSubmoduleDetailedStatusAsync(currentModule.GetTopModule(), cancelToken);
                 }
 
-                // Start update status for the submodules
-                if (updateStatus)
+                if (_gitStatusWhileUpdatingStructure != null)
                 {
-                    if (result.SuperProject != null)
-                    {
-                        // Update from top module (will stop at current)
-                        await GetSubmoduleDetailedStatusAsync(currentModule.GetTopModule(), cancelToken);
-                    }
-
-                    if (_gitStatusWhileUpdatingStructure != null)
-                    {
-                        // Current module must be updated separately (not in _submoduleInfos)
-                        await UpdateSubmodulesStatusAsync(currentModule, _gitStatusWhileUpdatingStructure, cancelToken);
-                    }
-
-                    OnStatusUpdated(result, cancelToken);
+                    // Current module must be updated separately (not in _submoduleInfos)
+                    await UpdateSubmodulesStatusAsync(currentModule, _gitStatusWhileUpdatingStructure, cancelToken);
                 }
 
-                _submoduleInfoResult = result;
-            }).FileAndForget();
+                OnStatusUpdated(result, cancelToken);
+            }
+
+            _submoduleInfoResult = result;
         }
 
         /// <inheritdoc />
-        public void UpdateSubmodulesStatus(string workingDirectory, [CanBeNull] IReadOnlyList<GitItemStatus> gitStatus, bool forceUpdate)
+        public async Task UpdateSubmodulesStatusAsync(string workingDirectory, [CanBeNull] IReadOnlyList<GitItemStatus> gitStatus, bool forceUpdate)
```

Comment:
When I implemented the 'pruning' of submodule updates I did not get the cancellation token working properly so I implemented dropping of frequent updates only. The variable handling was a concern when reviewing this.
Fixed this, could make a parallel PR but rather wait until until this is merged.

https://github.com/gerhardol/gitextensions/tree/feature/i7097-cleanup-status-update

Reply:
@gerhardol, could you have also a look a the disabled assertion in `Submodule_status_changes_for_first_nested_module` (#7443), please?

Reply:
I have looked at it, will continue later

Reply:
Submitted #8119 for the throttling changes
Tests are separate